### PR TITLE
Attempt at biginteger support

### PR DESCRIPTION
**Pull request by afk11**
I was trying to serialize an elliptic curve prime as an Integer with this library, but I noticed whenever I provide this (as a string - the result of gmp_strval), use of pure PHP math operators in the Integer class winds up casting the value as an integer, which is limited by PHP_INT_MAX. https://github.com/fgrosse/PHPASN1/blob/master/lib/ASN1/Universal/Integer.php#L62-71

Note that this problem of initializing with the wrong value doesn't happen when parsing structures from elsewhere, since using GMP allows for arbitrarily sized integers. https://github.com/fgrosse/PHPASN1/blob/master/lib/ASN1/Universal/Integer.php#L85-95

I have a sort of working attempt here and a test for it also. It's a shame that PHP doesn't support arbitrarily sized integers yet :/ 

Anyways, it would be nice to support writing these integers somehow!

### CODE REVIEW

**Comment by afk11** [comment]
Tests fail only because of StyleCI, which seems to complain about some of the phpdoc type hinting in numerous classes that I didn't actually change. For coherence I won't change these in this PR.

**Comment by fgrosse** [comment]
Thanks for the PR @afk11 I will probably look at this PR over the weekend.
From quickly glancing over your description I guess this will solve issue #1 right?

**Comment by afk11** [comment]
Hi @fgrosse - yes I believe this should fix it. That said, I would appreciate if you could try running 'afk11/asn1', a fork with this patch put through, in some of your applications to rigorously test? I am a bit unsure about the handling of large negative numbers. (see one of the tests). 

All I did was mirror the operations using GMP equivalents, but I worry I might have overlooked something - AFAIK the standards use two's complement for integers, making 2^256 a negative number when encountered in binary.. So if you have any tests that might already capture this, it'd be cool to see how it does.

File: lib/ASN1/Universal/Integer.php [hunk 1]
```diff
@@ -42,15 +42,20 @@ public function getContent()
         return $this->value;
     }
 
+    public function rightShift($number, $positions)
```
Comment (fgrosse) [inline lib/ASN1/Universal/Integer.php]: I there a reason why this method is public? Seems to me like this should be private.

File: lib/ASN1/Universal/Integer.php [hunk 2]
```diff
@@ -42,15 +42,20 @@ public function getContent()
...
+        return gmp_strval(gmp_div($number, gmp_pow(2, (int)$positions)));
+    }
+
     protected function calculateContentLength()
     {
         $nrOfOctets = 1; // we need at least one octet
-        $tmpValue = abs($this->value);
-        while ($tmpValue > 127) {
-            $tmpValue = $tmpValue >> 8;
+        $tmpValue = gmp_abs(gmp_init($this->value));
```
Comment (fgrosse) [inline lib/ASN1/Universal/Integer.php]: According to [the documentation][1] you don't need the `gmp_init` here.

> Note:
It is not necessary to call this function if you want to use integer or string in place of GMP number in GMP functions, like gmp_add(). Function arguments are automatically converted to GMP numbers, if such conversion is possible and needed, using the same rules as gmp_init().

[1]: http://php.net/manual/en/function.gmp-init.php

File: lib/ASN1/Universal/Integer.php [hunk 3]
```diff
@@ -60,14 +65,13 @@ protected function getEncodedValue()
         $contentLength = $this->getContentLength();
 
         if ($numericValue < 0) {
```
Comment (fgrosse) [inline lib/ASN1/Universal/Integer.php]: Will this check work if `$this->value` is a huge negative integer?

Maybe it would be better to initialize `$this->value` in the constructor and from then on always rely on it being a gmp value and consequently apply the according gmp operations. What do you think?

File: lib/ASN1/Universal/Integer.php (hunk 1 above)
Comment (afk11) [inline lib/ASN1/Universal/Integer.php]: No real reason, I can change that.

File: lib/ASN1/Universal/Integer.php (hunk 2 above)
Comment (afk11) [inline lib/ASN1/Universal/Integer.php]: Alright, that's cool!

File: lib/ASN1/Universal/Integer.php (hunk 3 above)
Comment (afk11) [inline lib/ASN1/Universal/Integer.php]: Hmm, I'll need to replace that with gmp_sign I suspect, but it should work then. I'll add some more tests for positive / negative numbers from 2^1 - 2^256

Main stuff to check is the twos complement stuff.. The first bit/byte is used a mask to indicate the sign. so AFAIK, given an n-bit integer, we can only save −(2^n−1) to +(2^(n−1) − 1).

Will send another commit for this today

File: lib/ASN1/Universal/Integer.php (hunk 2 above)
Comment (aztech-dev) [inline lib/ASN1/Universal/Integer.php]: Beware with automatic gmp conversions, some values can be misinterpreted as octal/binary. The safe way is to specify the base of the value with gmp_init...

Comment (afk11) [inline lib/ASN1/Universal/Integer.php]: Hmm good point @aztech-dev. I was worried about maybe dropping the sign, but it could end up being worse than that altogether - I will add the base and leave gmp_init in to be safe! 

One of the latest/scariest in a line of related problems: http://www.darkreading.com/vulnerabilities---threats/php-hash-comparison-weakness-a-threat-to-websites-researcher-says-/d/d-id/1320353

Comment (fgrosse) [inline lib/ASN1/Universal/Integer.php]: Good point indeed. @afk11 can you write a test for that so we are are sure this is covered and will not come back later as a regression?

**Comment by afk11** [comment]
I have been busy the past while, but will follow up with something today tomorrow

**Comment by fgrosse** [comment]
Yeah I have been distracted with other thing myself so dont worry. If you need some help drop a  line.

File: lib/ASN1/Universal/Integer.php (hunk 3 above)
Comment (afk11) [inline lib/ASN1/Universal/Integer.php]: It appears setting saving the result of gmp_init() in the constructor causes problems elsewhere in the tests. I'm using PHP5.5 at the moment, and since this has gmp resources, doing assertEquals fails. 

I might avoid this for now to keep this change simple.

**Comment by afk11** [comment]
I finished this up and addressed the nits from before. It has more tests for large negative/positive numbers, so think it's good to go.

**Comment by afk11** [comment]
Ping @fgrosse - made some progress with storing the gmp_init() right away in Integer. I tweaked Integer::fromBinary() to call `new static(gmp_strval())` in order to create a proper instance. 

The only issue is with <=PHP5.5 (IIRC), GMP encapsulates values with resources instead of objects. Unfortunately this means objects with resources cannot be directly compared. 

It's actually worse than it sounds, because if Set contains an Integer, you can't compare the whole set with another. 

The way around this is to:
* don't save the result of `gmp_init($arg)`, and make sure it's called every time `$this->value` is used throughout the code
* Don't directly compare $object with $parsedObject. 

I have an extreme distaste for not being able to compare objects directly - serialization logic shouldn't be necessary just for this, and the side effects for larger structures is not good. What do you think?

**Comment by fgrosse** [comment]
I agree that storing the gmp_init value is just not worth loosing comparability.
According to some benchmarks I ran later that day it doesnt even improve overall speed so we can savely leave it as it is now :)
